Passing parameters to can() via AccessRule

fsateler opened the pull request:
I am currently using this solution to make it easy to implement authorization based on query parameters.

For example, determining whether a post can be edited is determined by a Rule based on the post id. This implementation allows this type of use case to be easily done, without having to add a `matchCallback` to each action that needs a specific parameter to determine permission.

This is a first attempt at the problem, so I'm looking for feedback. If the feature is well received I can also modify the authorization section of the guide to explain usage.

Examples, given the following Rule:

``` php
    class EditOwnRule extends \yii\rbac\Rule {
        public $name = 'editOwnRule';

        public function execute($user, $item, $params) {
            if (Yii::$app->user->isGuest) {
                return false;
            }

            $item = ArrayHelper::getValue($params, 'item', null);
            if (!$item) {
                return false;
            }
            $q = Item::find()
                    ->where([
                        'owner_id' => $user,
                        'item.id' => $item
                    ]);
            return $q->exists();
        }
    }
```

Using this rule in an action would be as simple as doing the following:

``` php
    public function behaviors()
    {
        return [
            'access' => [
                'class' => AccessControl::className(),
                'rules' => [
                    [
                        'allow' => true,
                        'roles' => ['edit'],
                        'paramsMap' => [
                            // This will pass the 'id' query parameter as 'item' to can()
                            'id' => 'item',
                        ],
                        'actions' => [
                            'update',
                        ],
                    ],
                    [
                        'allow' => true,
                        'roles' => ['edit'],
                        'paramsMap' => [
                            // This will pass the 'item' query parameter as 'item' to can()
                            'item',
                        ],
                        'actions' => [
                            'updateComments',
                        ],
                    ],
                    [
                        'allow' => false,
                    ]
                ],
            ],
        ];
    }
```

fsateler committed:
Enh #8426: `yii\filters\AccessRule` now allows passing GET parameters to the role checking function

Updates:
	Add PR number and nick to changelog line
	Fixes suggested by @samdark
	Rename to paramsMap to queryParams, clarify phpdoc
	Otherwise queryParam and queryParams are too similar

lynicidn commented:
it's only for role check https://github.com/yiisoft/yii2/issues/6692

fsateler commented:
@lynicidn yes, only for role check. #7643 and #4302 are also related

samdark commented on framework/CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ Yii Framework 2 Change Log
 - Enh #8070: `yii\console\controllers\MessageController` now sorts created messages, even if there is no new one, while saving to PHP file (klimov-paul)
 - Enh #8286: `yii\console\controllers\MessageController` improved allowing extraction of nested translator calls (klimov-paul)
 - Enh #8415: `yii\helpers\Html` allows correct rendering of conditional comments containing `!IE` (salaros, klimov-paul)
+- Enh: `yii\filters\AccessRule` now allows passing GET parameters to the role checking function
```
Neeed issue number `#8426` and your nickname after the line.

fsateler replied:
Added.

fsateler commented:
BTW, I welcome alternative names for the `$paramsMap` property. It isn't quite as descriptive as one would like....

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+     * The special parameters '{action}' and '{controller}' are bound to the
+     * action and controller ids
+     */
+    public $paramsMap = [];
```
queryParameters?

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+    public $paramsMap = [];
+    
+    
+    private $_paramsForCan = null;
```
parametersForCan?

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+                if ($key == '{action}') {
+                    $newParams[$value] = Yii::$app->controller->action->id;
+                }
+                else if ($key == '{controller}') {
```
`elseif`

samdark replied:
} elseif {

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+     * The map indicates query parameters that will be passed when determining
+     * role access.
+     * 
+     * If only a value is passed then the query parameter will by passed as is.
```
Not clear what the sentence means.

fsateler replied:
Basically it means that passing only a value is equivalent to passing `value => value`. Not sure how to phrase it better

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+     * role access.
+     * 
+     * If only a value is passed then the query parameter will by passed as is.
+     * If both a key and value are passed, then the name of the parameter will
```
a key and **a** value

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+     * 
+     * If only a value is passed then the query parameter will by passed as is.
+     * If both a key and value are passed, then the name of the parameter will
+     * be changed. A value of `['a' => 'b']` will invoke [[User::can()]] with
```
will invoke X → will cause X to be invoked

samdark commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,47 @@ class AccessRule extends Component
+    
+    
+    private $_paramsForCan = null;
+    private function getParamsForCan() {
```
Need PHPdoc for private stuff as well.

samdark commented:
It's a good idea to mention it in the guide.

fsateler commented:
@samdark Fixed most of your comments. I'm holding on the variable name change to see if others have other preference.

I will add a section to the guide later.

cebe commented:
> // This will pass the 'id' query parameter as 'item' to can()

whats the benefit of this? you could just get the parameter inside of can() then...

fsateler commented:
@cebe because neither can() or the Rule have enough information to determine which query parameter should be used. The same rule could be used in one controller where 'id' is the query parameter to look at, and other controller where it should look at 'item_id'

fsateler committed:
Document queryParams in the authorization guide

fsateler commented:
I have rebased on top of master, resolved conflicts and added a short section on the guide. Comments welcome.

I also renamed to queryParams since in the end that is all we are looking at, so it is a good name.

PommeVerte commented:
I'm glad this is getting some attention. Keeps access rules clear and readable. Been using lambdas so far and it's been quite verbose.

fsateler commented:
I have rebased on top of current master and fixed the merge errors.

PowerGamer1 commented:
Just stumbled across the same problem discussed here and I don't like the proposed solution at all because it has quiet significant flaw: the model gets loaded from database TWICE - once in the implementation of RBAC rule (to perform access check) and the second time in the controller action itself (when the access to it was granted and it proceeds to perform its duty). Making TWO queries into database for exactly the SAME data is BAD from performance and data integrity points of view.

For the moment the only solution I see is to perform an access check inside a controller action itself and forfeit the use of built-in AccessControl filter entirely.

fsateler commented:
@PowerGamer1 that would depend on how do you implement the queries. You
could use caching.

But more importantly, it is rarely the case that time spent in the database
is more important than developer time.
On 25 Jun 2015 07:39, "PowerGamer1" <notifications@github.com> wrote:

> Just stumbled across the same problem discussed here and I don't like the
> proposed solution at all because it has quiet significant flaw: the model
> gets loaded from database TWICE - once in the implementation of RBAC rule
> (to perform access check) and the second time in the controller action
> itself (when the access to it was granted and it proceeds to perform its
> duty). Making TWO queries into database for exactly the SAME data is BAD
> from performance and data integrity points of view.
>
> For the moment the only solution I see is to perform an access check
> inside a controller action itself and forfeit the use of built-in
> AccessControl filter entirely.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yiisoft/yii2/pull/8426#issuecomment-115206405>.
>

fsateler commented:
Ping. Is this change wanted? If so I can rebase and fix conflicts.

PommeVerte commented:
I dont want to start a +1 flood, but it would be nice to see this as the current implementation gets pretty verbose pretty quickly.

lynicidn commented:
i think better way: 
```php
'roles' => [$roleName => $roleParams]
```

php5developer commented:
I fully agree with **PowerGamer1**, we need to do something with query optimization.

fsateler commented:
ping. I can do a rebase and fix errors if this change is wanted. Otherwise lets close this.

fsateler commented:
@cebe I see you have assigned this issue. Does this mean there is interest in merging? If so, I can update the PR.

cebe commented:
not yet, I think there must be some changes to the approach but I had no time to review it in detail yet.

fsateler commented:
OK, just let me know if you need me to change the PR.

On 1 December 2016 at 17:22, Carsten Brandt <notifications@github.com>
wrote:

> not yet, I think there must be some changes to the approach but I had no
> time to review it in detail yet.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yiisoft/yii2/pull/8426#issuecomment-264283869>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABQsHcRYz0ImAvFJbSxE8BBxNbEaenzTks5rDyyigaJpZM4EbG5g>
> .
>



-- 

Saludos,
Felipe Sateler

Faryshta commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,62 @@ class AccessRule extends Component
+     * If only a value is passed then it is equivalent to passing the same value
+     * as key.
+     * 
+     * The special parameters '{action}' and '{controller}' are bound to the
```
i think `{route}` would be better since i can have `api/v1/user/index` and `api/v2/user/index` with different permissions.

Faryshta commented on framework/filters/AccessRule.php:
```diff
@@ -90,6 +92,62 @@ class AccessRule extends Component
     public $denyCallback;
 
+    /** 
+     * @var array a map for passing parameters to the [[User::can()]] function
```
how about allowing for callbacks? this would ease the implementation a lot.

'params' => function ($action, $accessRule) {
     return ['autor_id' => Yii::$app->user->id];
}

cebe replied:
see #14123 for an alternative implementation using closures.

cebe commented:
see #14123 for an alternative implementation using closures.

cebe commented:
closing in favor of  #14123, thanks for you proposal!
